### PR TITLE
Enable an I/O (FieldSet) caching

### PR DESCRIPTION
**Pull request by CKehl**
Locally as well as on cluster, user may have - next to the original storage location of the fieldset files - another storage location for cached, high-speed file access (e.g. a solid-state drive). The idea, logically, would be to automatically use this high-speed storage location as cache to pre-buffer fieldset data and thus increase performance.

The initial motivation came from the OceanParcels internal, brand-new LORENZ cluster, which has high-speed I/O facilities for this purpuse. That said, the caching class is ideally made so that people can setup this caching mode also for their cluster, or even for a local laptop with multiple SSD's.

### CODE REVIEW

File: parcels/fieldset.py [hunk 1]
```diff
@@ -83,8 +95,10 @@ def from_data(cls, data, dimensions, transpose=False, mesh='spherical',
         :param allow_time_extrapolation: boolean whether to allow for extrapolation
                (i.e. beyond the last available time snapshot)
                Default is False if dimensions includes time, else True
-        :param time_periodic: To loop periodically over the time component of the Field. It is set to either False or the length of the period (either float in seconds or datetime.timedelta object). (Default: False)
-               This flag overrides the allow_time_interpolation and sets it to False
+        :param To loop periodically over the time component of the Field. It is set to either :param False or
```
Comment (CKehl) [inline parcels/fieldset.py:98]:
```suggestion
        :param time_periodic: To loop periodically over the time component of the Field. It is set to either :param False or
```

File: parcels/fieldset.py [hunk 2]
```diff
@@ -343,6 +363,10 @@ def from_netcdf(cls, filenames, variables, dimensions, indices=None, fieldtype=N
                '{parcels_varname: {netcdf_dimname : (parcels_dimname, chunksize_as_int)}, ...}', where 'parcels_dimname' is one of ('time', 'depth', 'lat', 'lon')
         :param netcdf_engine: engine to use for netcdf reading in xarray. Default is 'netcdf',
                but in cases where this doesn't work, setting netcdf_engine='scipy' could help
+        :param cache: a custom FieldFileCache object, or None for automatic construction (default: None)
+        :param cache_dir: auto-constructing a FieldFileCache object with this given top-leave path as cache directory (default: None)
+        :param do_cache: switch to enable or disable caching (default: True [= enabled])
```
Comment (CKehl) [inline parcels/fieldset.py:368]:
```suggestion
        :param do_cache: switch to enable or disable caching (default: False [= disabled])
```

File: parcels/fieldset.py [hunk 3]
```diff
@@ -746,11 +794,17 @@ def from_mom5(cls, filenames, variables, dimensions, indices=None, mesh='spheric
...
+        :param time_periodic: To loop periodically over the time component of the Field. It is set to either :param False or
+               the total timespan that is to be covered by the periodically-repeating field, given either as :type float (in seconds)
+               or :type datetime.timedelta object). (Default: False)
+               This flag overrides the allow_time_interpolation and sets it to False.
         :param tracer_interp_method: Method for interpolation of tracer fields. It is recommended to use 'bgrid_tracer' (default)
                Note that in the case of from_mom5() and from_bgrid(), the velocity fields are default to 'bgrid_velocity'
         :param chunksize: size of the chunks in dask loading
+        :param cache: a custom FieldFileCache object, or None for automatic construction (default: None)
+        :param cache_dir: auto-constructing a FieldFileCache object with this given top-leave path as cache directory (default: None)
+        :param do_cache: switch to enable or disable caching (default: True [= enabled])
```
Comment (CKehl) [inline parcels/fieldset.py:806]:
```suggestion
        :param do_cache: switch to enable or disable caching (default: False [= disabled])
```

File: parcels/fieldset.py [hunk 4]
```diff
@@ -863,13 +923,19 @@ def from_parcels(cls, basename, uvar='vozocrtx', vvar='vomecrty', indices=None,
...
+               or :type datetime.timedelta object). (Default: False)
+               This flag overrides the allow_time_interpolation and sets it to False.
         :param deferred_load: boolean whether to only pre-load data (in deferred mode) or
                fully load them (default: True). It is advised to deferred load the data, since in
                that case Parcels deals with a better memory management during particle set execution.
                deferred_load=False is however sometimes necessary for plotting the fields.
         :param chunksize: size of the chunks in dask loading
+        :param cache: a custom FieldFileCache object, or None for automatic construction (default: None)
+        :param cache_dir: auto-constructing a FieldFileCache object with this given top-leave path as cache directory (default: None)
+        :param do_cache: switch to enable or disable caching (default: True [= enabled])
```
Comment (CKehl) [inline parcels/fieldset.py:937]:
```suggestion
        :param do_cache: switch to enable or disable caching (default: False [= disabled])
```

File: parcels/fieldset.py [hunk 5]
```diff
@@ -479,14 +513,18 @@ def from_nemo(cls, filenames, variables, dimensions, indices=None, mesh='spheric
...
+               the total timespan that is to be covered by the periodically-repeating field, given either as :type float (in seconds)
+               or :type datetime.timedelta object). (Default: False)
+               This flag overrides the allow_time_interpolation and sets it to False.
         :param tracer_interp_method: Method for interpolation of tracer fields. It is recommended to use 'cgrid_tracer' (default)
                Note that in the case of from_nemo() and from_cgrid(), the velocity fields are default to 'cgrid_velocity'
         :param chunksize: size of the chunks in dask loading. Default is None (no chunking)
-
+        :param cache: a custom FieldFileCache object, or None for automatic construction (default: None)
+        :param cache_dir: auto-constructing a FieldFileCache object with this given top-leave path as cache directory (default: None)
+        :param do_cache: switch to enable or disable caching (default: True [= enabled])
```
Comment (CKehl) [inline parcels/fieldset.py:525]:
```suggestion
        :param do_cache: switch to enable or disable caching (default: False [= disabled])
```

**Comment by CKehl** [comment]
@erikvansebille Summary of the PR:

Purpose: buffer (or: cache) Field files on  e.g. SSD's locally and on clusters -> use the SSD's on LORENZ nodes
Parent branch: **master** (3.x and onwards) -> needs to merge in changes of _zarr_ if need be
Status: merge-ready from momentary **master** head and _can_ be merged to master -> changes are rather minor; tests available; tutorial notebook available; most business is going on in _filecache.py_
Documentation status: test is available; tutorial notebook is available; discussed the process and algorithm with @lvankampenhout in detail in June/July

advice: please merge into **master** after tests finish

**Comment by erikvansebille** [comment]
Closing this PR for now, as will not integrate into master anytime soon

**Pull request closed without merging**